### PR TITLE
NDRS-620: Publish evidence of an equivocation to the SSE Server event stream.

### PR DESCRIPTION
**Pull request by darthsiroftardis**
REF: https://casperlabs.atlassian.net/browse/NDRS-620

Sample Output:
```
{
   "EquivocationEvent":{
      "era_id":0,
      "evidence":{
         "creator_id":4,
         "creator_public_key":"020248509e67db3127f82d5224c5c18eac00f96d1edeadbadc8eb2c8606227b56873",
         "instance":"7014550edb2f4c96db3c2f1c9789b107c8998749326a5f98dcb126668426ecac",
         "sequence_number":0,
         "round_exp":[
            12,
            12
         ],
         "timestamps":[
            "2020-12-07T15:50:56.269Z",
            "2020-12-07T15:50:56.270Z"
         ],
         "signatures":[
            "0284bb6868c8920f2b3af45485212a7befa9d905b90527a1212e04eaf95a09551e4211841b2fb8fc8dfa5a127dea3e2d54e38a2a88434da53d3655e7a9340a8088",
            "02ca2184ec08cde81ac7ec5dc455cd1ae1473a5dca22e191f12711fe85daf731e009f7de029bf3e311605869e9c07efc13259924d398644c33719985ac98964948"
         ]
      }
   }
}
```

### CODE REVIEW

**Comment by goral09** [comment]
Round exponents will be the same for both units.

**Comment by afck** [comment]
> Round exponents will be the same for both units.

Not necessarily. Not sure whether it's useful to put them in the event, though.

File: node/src/components/consensus/consensus_protocol.rs [hunk 1]
```diff
@@ -85,6 +92,8 @@ pub(crate) enum ProtocolOutcome<I, C: Context> {
     NewEvidence(C::ValidatorId),
     /// Send evidence about the validator from an earlier era to the peer.
     SendEvidence(I, C::ValidatorId),
+    /// Send evidence about an equivocation to the event stream.
+    EquivocationEvent(Evidence<C>),
```
Comment (afck) [inline node/src/components/consensus/consensus_protocol.rs:96]: We shouldn't leak Highway details; that breaks pluggability for consensus.

**Review by afck — CHANGES_REQUESTED** [review]
Let's not leak all those Highway implementation details. We need to find another way to get information about an equivocation into the event stream.

Maybe we could add some trait for consensus protocol faults, and then `Box` that inside of `HighwayProtocol`?

File: node/src/components/consensus/consensus_protocol.rs (hunk 1 above)
Comment (darthsiroftardis) [reply node/src/components/consensus/consensus_protocol.rs:96]: `Evidence<C>` is sent to the `era_supervisor` and is then mapped onto an external type `EvidenceStream` to be sent to the event stream. Would that still classify as leaking details?

Comment (afck) [reply node/src/components/consensus/consensus_protocol.rs:96]: Yes, `Evidence` is Highway-specific.

Comment (xcthulhu) [reply node/src/components/consensus/consensus_protocol.rs:96]: @afck Maybe have the context `C` have an associated `C::Evidence` ?

**Comment by xcthulhu** [comment]
What is the testing strategy for this?

File: node/src/components/consensus/consensus_protocol.rs (hunk 1 above)
Comment (afck) [reply node/src/components/consensus/consensus_protocol.rs:96]: `Context` isn't implemented by the protocol, but… by all the rest! The `ClContext` specifies which types the Casper node is using, even though `Highway` could work with other contexts (and does, for faster and simpler tests).

Anyway, we decided to just push non-Highway-specific information on the event stream anyway; @darthsiroftardis will create a new PR that should be simpler.

**Pull request closed without merging**